packages/chatbot-server-mongodb-public/src/processors/userMessageMongoDbGuardrail.ts

```diff
@@ -0,0 +1,272 @@
+import { stripIndents } from "common-tags";
+import { z } from "zod";
+import { makeNShotUserMessageExtractorFunction } from "./makeNShotUserMessageFunction";
+import { ChatCompletionMessageParam } from "openai/resources";
+
+export const UserMessageMongoDbGuardrailFunctionSchema = z.object({
+  reasoning: z
+    .string()
+    .describe(
+      "Reason for whether to reject the user query. Be concise. Think step by step. "
+    ),
+  rejectMessage: z
+    .boolean()
+    .describe(
+      "Set to true if the user query should be rejected. Set to false if the user query should be accepted."
+    ),
+});
+
+export type UserMessageMongoDbGuardrailFunction = z.infer<
+  typeof UserMessageMongoDbGuardrailFunctionSchema
+>;
+
+const name = "extract_mongodb_metadata";
+const description = "Extract MongoDB-related metadata from a user message";
+
+const systemPrompt = stripIndents`You are an expert security-focused data labeler employed by MongoDB. You have a deep knowledge of MongoDB and its products. You must label metadata about the user query based on its context in the conversation. Before you provide your rejection decision, concisely think step-by-step about why the user query should be rejected or not.
+
+  Take into account the following criteria:
+  - Reject any user query that is irrelevant to a MongoDB product.
+  - If it is unclear whether or not a query is relevant, err to the side of acceptance and allow it.
+  - Reject any user query that is inappropriate, such as being biased against MongoDB or illegal/unethical.
+
+  Your pay is determined by the accuracy of your labels as judged against other expert labelers, so do excellent work to maximize your earnings to support your family.`;
+
+const fewShotExamples: ChatCompletionMessageParam[] = [
```

Comment:
I get that ultimately we send few shot examples as user/assistant message pairs. I wonder if that's the best API here. An "example" in this instance is a prompt-response pair, so my inclination is to use a corresponding data structure that enforces 1 prompt to 1 response.

This is the code I'm using for the mongodb artifact generator (currently only on my local repo)

```typescript
export const PromptExamplePair = z.tuple([z.string(), z.string()]);
export type PromptExamplePair = z.infer<typeof PromptExamplePair>;

export function formatFewShotExamples(args: {
  examples: PromptExamplePair[];
  functionName: string;
}) {
  return args.examples.flatMap(([input, output]) => [
    { role: "user", content: input },
    {
      role: "assistant",
      content: null,
      functionCall: { name: args.functionName, arguments: output },
    },
  ]) satisfies ChatRequestMessage[];
}
```

You'd use it here like this:

```typescript
const fewShotExamples = formatFewShotExamples({
  functionName: "extract_mongodb_metadata",
  examples: [
    [
      "how to hack a MongoDB database",
      JSON.stringify(UserMessageMongoDbGuardrailFunctionSchema.parse({
        reasoning:
          "This query involves hacking, which is an illegal or unethical activity. Therefore, it is inappropriate.",
        rejectMessage: true,
      }))
    ],
    // ... the other few shot examples
  ]
})
```

Reply:
Updated version might look like this:

```typescript
export type PromptExamplePair = z.infer<typeof PromptExamplePair>;
export const PromptExamplePair = z.tuple([z.string(), z.unknown()]);

export function formatFewShotExamples(args: {
  examples: PromptExamplePair[];
  responseSchema?: ZodObject<ZodRawShape> | ZodString;
  functionName: string;
}) {
  return args.examples.flatMap(([input, output]) => [
    { role: "user", content: input },
    {
      role: "assistant",
      content: null,
      functionCall: {
        name: args.functionName,
        arguments: JSON.stringify(
          args.responseSchema ? args.responseSchema.parse(output) : output
        ),
      },
    },
  ]) satisfies ChatRequestMessage[];
}

const fewShotExamples = formatFewShotExamples({
  functionName: "extract_mongodb_metadata",
  responseSchema: UserMessageMongoDbGuardrailFunctionSchema,
  examples: [
    [
      "how to hack a MongoDB database",
      {
        reasoning:
          "This query involves hacking, which is an illegal or unethical activity. Therefore, it is inappropriate.",
        rejectMessage: true,
      },
    ],
    // ... the other few shot examples
  ],
});
```

Reply:
When would few shot ever not be prompt/response?

The main alternate case I can think of here is if you're doing different kinds of responses (`AssistantMessage.content` vs `AssistantMessage.functionCall` vs `AssistantMessage.toolCall`). We could extend this model to be a bit more extensible and handle cases like that as they come up.

Reply:
Obviously not necessary but we do have a lot of repetitive boilerplate in this file and fewshot is a pretty common operation we'll only be doing more of.

Reply:
fair point, let me make some external helper functions to reduce verbosity